dlt/transformations/transformation.py

```diff
@@ -66,130 +94,110 @@ def make_transformation_resource(
-                hints=make_hints(columns=all_columns),
-            )
-        else:
-            # NOTE: dataset will not execute query over unknown tables or columns
-            for chunk in datasets[0](select_query).iter_arrow(chunk_size=config.buffer_max_items):
-                yield dlt.mark.with_hints(chunk, hints=make_hints(columns=all_columns))
+                should_materialize = False
+        # respect config setting
+        should_materialize = should_materialize or config.always_materialize
+
+        # add check to only allow one transformation to be processed per function
+        transformation_processed = False
+
+        def _process_item(item: TDataItems) -> Iterator[TDataItems]:
+            # catch the cases where we get a relation from the transformation function
+            if isinstance(item, ReadableDBAPIRelation):
+                relation = item
+            # we see if the string is a valid sql query, if so we need a dataset
+            elif isinstance(item, str):
+                try:
+                    sqlglot.parse_one(item)
+                    if len(datasets) == 0:
+                        raise IncompatibleDatasetsException(
+                            resource_name,
+                            "No datasets found in transformation function arguments. Please supply"
+                            " all used datasets via transform function arguments.",
+                        )
+                    else:
+                        relation = datasets[0](item)
+                except sqlglot.errors.ParseError as e:
+                    raise TransformationException(
+                        resource_name,
+                        "Invalid SQL query in transformation function. Please supply a valid SQL"
+                        " query via transform function arguments.",
+                    ) from e
+            elif IbisExpr and isinstance(item, IbisExpr):
+                relation = datasets[0](item)
+            else:
+                # no transformation, just yield this item
+                yield item
+                return
+
+            nonlocal transformation_processed
+            transformation_processed = True
+
+            if not should_materialize:
+                yield relation
+            else:
+                from dlt.common.libs.pyarrow import add_arrow_metadata
+
+                serialized_hints = json.dumps(relation.compute_hints())
+                for chunk in relation.iter_arrow(chunk_size=config.buffer_max_items):
+                    yield add_arrow_metadata(chunk, {DLT_HINTS_METADATA_KEY: serialized_hints})
+
+        # support both generator and function
+        gen_or_item = func(*args, **kwargs)
+        iteratable_items = gen_or_item if isinstance(gen_or_item, Iterator) else [gen_or_item]
+
+        for item in iteratable_items:
+            if transformation_processed:
+                raise TransformationException(
```

Comment:
heh why we should block it? maybe single query per transformation is a good thing but I can totally see people yielding model and a few arrow tables with intermediary data. or dispatch models to tables via dynamic table names. so they will yield many models.

ie. a model per each type of event in github

Reply:
this was blocked because the hints did not update and I did not know why, now I do and this works and I allow it an uncommented the test :)